File: 3633-maximize-the-number-of-target-nodes-after-connecting-trees-i/maximize-the-number-of-target-nodes-after-connecting-trees-i.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    vector<int> maxTargetNodes(vector<vector<int>>& edges1,
                               vector<vector<int>>& edges2,
                               int k) {
        int n = edges1.size() + 1;
        int m = edges2.size() + 1;

        // build adjacency lists
        vector<vector<int>> adj1(n), adj2(m);
        for (auto &e : edges1) {
            int u = e[0], v = e[1];
            adj1[u].push_back(v);
            adj1[v].push_back(u);
        }
        for (auto &e : edges2) {
            int u = e[0], v = e[1];
            adj2[u].push_back(v);
            adj2[v].push_back(u);
        }

        // g1[i] = # nodes in tree1 within distance ≤ k of i
        // g2[j] = # nodes in tree2 within distance ≤ k−1 of j
        auto g1 = computeWithinDist(n, adj1, k);
        auto g2 = computeWithinDist(m, adj2, k-1);

        // best you can do in tree2 by choosing the best attach-point
        int M = 0;
        if (k > 0) 
            M = *max_element(g2.begin(), g2.end());

        // answer[i] = g1[i] + M
        vector<int> answer(n);
        for (int i = 0; i < n; i++) {
            answer[i] = g1[i] + M;
        }
        return answer;
    }

private:
    struct Info { int c, dist, comp; };

    // Returns for each u in [0..N-1], the count of nodes within distance ≤ R of u.
    // Runs in O(N log²N) via centroid decomposition.
    vector<int> computeWithinDist(int N,
                                  const vector<vector<int>>& adj,
                                  int R) {
        vector<bool> blocked(N, false);
        vector<int>  subSize(N, 0), centroidParent(N, -1);
        vector<vector<Info>>                  layers(N);
        vector<vector<int>>                   allDist(N);
        vector<vector<vector<int>>>           childDist(N);

        // 1) compute subtree sizes
        auto dfsSize = [&](auto&& self, int u, int p) -> void {
            subSize[u] = 1;
            for (int v : adj[u]) if (!blocked[v] && v != p) {
                self(self, v, u);
                subSize[u] += subSize[v];
            }
        };

        // 2) find centroid of a component of size sz rooted at u
        auto findCentroid = [&](auto&& self, int u, int p, int sz) -> int {
            for (int v : adj[u]) if (!blocked[v] && v != p
                                     && subSize[v] * 2 > sz)
                return self(self, v, u, sz);
            return u;
        };

        // 3) gather distances from centroid c into its component
        auto dfsDist = [&](auto&& self,
                           int u, int p, int d, int c, int comp) -> void {
            layers[u].push_back({c, d, comp});
            allDist[c].push_back(d);
            if (comp >= 0)
                childDist[c][comp].push_back(d);
            for (int v : adj[u]) if (!blocked[v] && v != p) {
                self(self, v, u, d + 1, c, comp);
            }
        };

        // 4) build the centroid‐decomposition
        auto decompose = [&](auto&& self, int entry, int p) -> void {
            dfsSize(dfsSize, entry, -1);
            int c = findCentroid(findCentroid, entry, -1, subSize[entry]);

            blocked[c]         = true;
            centroidParent[c]  = p;

            // record centroid -> itself
            layers[c].push_back({c, 0, -1});
            allDist[c].push_back(0);

            // count how many child‐subtrees remain
            int comps = 0;
            for (int v : adj[c]) if (!blocked[v]) comps++;
            childDist[c].assign(comps, {});

            // gather distances in each child‐subtree
            int idx = 0;
            for (int v : adj[c]) if (!blocked[v]) {
                dfsDist(dfsDist, v, c, 1, c, idx);
                idx++;
            }

            // sort for binary‐search
            sort(allDist[c].begin(), allDist[c].end());
            for (int i = 0; i < comps; i++)
                sort(childDist[c][i].begin(), childDist[c][i].end());

            // recurse on each subtree
            for (int v : adj[c]) if (!blocked[v])
                self(self, v, c);
        };

        // start decomposition from node 0
        decompose(decompose, 0, -1);

        // now answer the “ball of radius R” query for every u
        vector<int> cnt(N, 0);
        for (int u = 0; u < N; u++) {
            long long reached = 0;
            for (auto &info : layers[u]) {
                int c    = info.c;
                int du   = info.dist;
                int comp = info.comp;
                int rem  = R - du;
                if (rem < 0) continue;

                // total in centroid c’s component within rem
                auto &A = allDist[c];
                long long tot = upper_bound(A.begin(), A.end(), rem)
                              - A.begin();

                // subtract the child‐subtree that contains u
                if (comp >= 0) {
                    auto &B = childDist[c][comp];
                    tot -= upper_bound(B.begin(), B.end(), rem)
                         - B.begin();
                }
                reached += tot;
            }
            cnt[u] = (int)reached;
        }

        return cnt;
    }
};

File: 3633-maximize-the-number-of-target-nodes-after-connecting-trees-i/maximize-the-number-of-target-nodes-after-connecting-trees-i.py
from bisect import bisect_right
from typing import List, Tuple

class Solution:
    def maxTargetNodes(self, edges1: List[List[int]],
                       edges2: List[List[int]],
                       k: int) -> List[int]:
        # build adjacency lists
        n = len(edges1) + 1
        m = len(edges2) + 1
        adj1 = [[] for _ in range(n)]
        adj2 = [[] for _ in range(m)]
        for u, v in edges1:
            adj1[u].append(v)
            adj1[v].append(u)
        for u, v in edges2:
            adj2[u].append(v)
            adj2[v].append(u)

        # g1[i] = # nodes in tree1 within distance ≤ k of i
        # g2[j] = # nodes in tree2 within distance ≤ k-1 of j
        g1 = self._withinDist(adj1, k)
        g2 = self._withinDist(adj2, max(0, k-1))
        M  = max(g2) if k > 0 else 0

        # answer[i] = g1[i] + M
        return [g1[i] + M for i in range(n)]

    def _withinDist(self,
                    adj: List[List[int]],
                    R: int
                   ) -> List[int]:
        """
        For each node u, computes the number of nodes within distance ≤ R
        in the tree given by adj, using centroid decomposition in O(N log^2 N).
        """
        self.adj       = adj
        self.R         = R
        N              = len(adj)
        self.blocked   = [False]*N
        self.subSize   = [0]*N
        self.layers    = [[] for _ in range(N)]      # List[List[Tuple[c,dist,comp]]]
        self.allDist   = [[] for _ in range(N)]      # distances from each centroid
        self.childDist = [[] for _ in range(N)]      # per-centroid child-subtree dists

        self._decompose(0)

        # Now answer for each u by walking its centroid-layers
        cnt = [0]*N
        for u in range(N):
            total = 0
            for c, dist, comp in self.layers[u]:
                rem = R - dist
                if rem < 0:
                    continue
                A = self.allDist[c]
                t = bisect_right(A, rem)
                if comp >= 0:
                    B = self.childDist[c][comp]
                    t -= bisect_right(B, rem)
                total += t
            cnt[u] = total
        return cnt

    def _dfsSize(self, u: int, p: int) -> int:
        """Compute subtree sizes ignoring blocked nodes."""
        self.subSize[u] = 1
        for v in self.adj[u]:
            if v != p and not self.blocked[v]:
                self.subSize[u] += self._dfsSize(v, u)
        return self.subSize[u]

    def _findCentroid(self, u: int, p: int, size: int) -> int:
        """Find centroid of the component rooted at u of total size `size`."""
        for v in self.adj[u]:
            if v != p and not self.blocked[v] and self.subSize[v]*2 > size:
                return self._findCentroid(v, u, size)
        return u

    def _dfsDist(self,
                 u: int,
                 p: int,
                 d: int,
                 c: int,
                 comp: int
                ) -> None:
        """
        From centroid c, record distance d for node u (in child-subtree comp)
        and recurse down its unblocked children.
        """
        self.layers[u].append((c, d, comp))
        self.allDist[c].append(d)
        if comp >= 0:
            self.childDist[c][comp].append(d)
        for v in self.adj[u]:
            if v != p and not self.blocked[v]:
                self._dfsDist(v, u, d+1, c, comp)

    def _decompose(self, entry: int) -> None:
        """Recursively build the centroid decomposition from `entry`."""
        size = self._dfsSize(entry, -1)
        c    = self._findCentroid(entry, -1, size)
        self.blocked[c] = True

        # 1) record the centroid itself
        self.layers[c].append((c, 0, -1))
        self.allDist[c].append(0)

        # 2) count how many child-subtrees
        comps = sum(1 for v in self.adj[c] if not self.blocked[v])
        self.childDist[c] = [[] for _ in range(comps)]

        # 3) gather distances in each child-subtree
        idx = 0
        for v in self.adj[c]:
            if not self.blocked[v]:
                self._dfsDist(v, c, 1, c, idx)
                idx += 1

        # 4) sort the collected distances for binary search
        self.allDist[c].sort()
        for sub in self.childDist[c]:
            sub.sort()

        # 5) recurse on each remaining component
        for v in self.adj[c]:
            if not self.blocked[v]:
                self._decompose(v)
